File: CSharp/FractionPerformance.cs
using System;
using System.Collections;
using System.Diagnostics;
using System.Runtime.InteropServices;


public struct Frequency
{
  public int _value;
  public int _count;

  public int Value() { return _value; }
  public int Count() { return _count; }
}

public class FrequencyArray
{
  Frequency [] _frequencies;
  int _capacity;
  int _size;
  int _maxFreq;

  class FrequencyComparer : IComparer
  {
    int IComparer.Compare(Object a,Object b)
    {
      return ((Frequency)a)._value - ((Frequency)b)._value;
    }
  }
  public FrequencyArray() { _capacity=16; _size=0; _maxFreq=0; _frequencies = new Frequency[_capacity]; }

  private int find(int v)
  {
    int i=0;
    for(;i<_size;i++)
      if(v==_frequencies[i]._value)
        break;
    return i;
  }

  public void Increment(int v) {
    int pos = find(v);
    if(pos >= _size) {
      _size++;
      if(pos >= _capacity) {
        _capacity<<=1;
        System.Array.Resize(ref _frequencies,_capacity);
      }
      _frequencies[pos]._value=v;
      _frequencies[pos]._count=0;
    }
    _frequencies[pos]._count++;
    if(_frequencies[pos]._count > _maxFreq)
      _maxFreq = _frequencies[pos]._count;
  }

  public void Sort()
  {
    Array.Sort(_frequencies,0,_size,new FrequencyComparer());
  }

  public int Length { get { return _size; } }
  public int MaxFreq { get { return _maxFreq; } }

  public int Value(int i) { return _frequencies[i]._value; }
  public int Frequency(int i) { return _frequencies[i]._count; }
}


/*
  LowValueFrequencyArray
  The frequency is not low, it is the value for which the frequency is counted that is low.
  Expected values are so low that the values are used as indexes.
*/
/*
struct LowValueFrequencyArray {
  // Increase size of array if an exception is thrown
  int [] _frequencies; // = new int[] { 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
  int _maxFreq;
  public LowValueFrequencyArray(int n)
  {
    _maxFreq=0;
    _frequencies = new int[n];
    for(int i=0;i<n;i++) _frequencies[i]=0;
  }
  public void Increment(int i)
  {
    if(i>_frequencies.Length)
      Console.WriteLine(String.Format("I ({0}) > length ({1})",i,_frequencies.Length));
    _frequencies[i]++;
    if(_frequencies[i]>MaxFreq)
      _maxFreq=_frequencies[i];
  }

  public int Frequency(int i) { return _frequencies[i]; }
  public int MaxFreq { get { return _maxFreq; }}
  public int Length { get { return _maxFreq+1; }}
}
*/
class Stats {
  private double _average;
  private double _standard_deviation;
  private int _size;
  private int _median;
  private int _mode;

  public void Calc(FrequencyArray freq_data)
  {
    // Average
    _size=0;
    int i,sum=0,max_freq=0;
    for(i=0;i<freq_data.Length;i++) {
      _size+=freq_data.Frequency(i);
      sum += i*freq_data.Frequency(i);
      if(freq_data.Frequency(i) > max_freq) {
        max_freq=freq_data.Frequency(i);
        _mode=i;
      }
    }
    _average = (double)sum/(double)_size;

    // median and variance
    double var=0;
    int count=0;
    _median=-1;
    for(i=0;i<freq_data.Length;i++) {
      var += (i - _average)*(i - _average)*freq_data.Frequency(i);
      count += freq_data.Frequency(i);
      if(_median == -1 && count >= _size/2) {
        _median = i;
      }
    }

    // standard deviation
    _standard_deviation=Math.Sqrt(var/(double)(_size-1));
  }

  public Fraction Average()
  {
    Fraction f=new Fraction(_average);
    f.Round(100);
    return f;
  }

  public Fraction StandardDeviation()
  {
    Fraction f = new Fraction(_standard_deviation);
    f.Round(100);
    return f;
  }

  public int Size() { return _size; }
  public int Median() { return _median; }
  public int Mode() { return _mode; }
}

class FractionPerformance
{
  const int MaxTicks = 20;  // The maximum clock ticke expected to calculate fraction from double
  const int MaxLoops = 10;  // The maximum number of loops to calculate fraction from double
  const int teminal_cols = 50;
  static void display_graph(FrequencyArray freq_data,string xlabel,string ylabel)
  {
    // Determine scale for graph
    int i;
    double scale = (double)teminal_cols/(double)freq_data.MaxFreq;
    Console.WriteLine("\n{0}|                 {1}",xlabel,ylabel);
    Console.WriteLine(new String('-',teminal_cols+6));
    for(i=0;i<freq_data.Length;i++) {
      string bar = new String('#',(int)Math.Round(freq_data.Frequency(i)*scale));
      Console.WriteLine("{0}|{1} {2}",freq_data.Value(i),bar,freq_data.Frequency(i));
    }
    Console.WriteLine();
  }

  static void ShowResults(FrequencyArray tick_freq,FrequencyArray loop_freq)
  {
    Console.WriteLine("Max time (in clock ticks): ",tick_freq.MaxFreq);
    Stats stats=new Stats();
    stats.Calc(tick_freq);
    Console.WriteLine("Sample size: "+stats.Size());
    Console.WriteLine("Average: "+stats.Average().ToStringMixed());
    Console.WriteLine("Median: "+stats.Median());
    Console.WriteLine("Mode: "+stats.Mode());
    Console.WriteLine("Standard Deviation: "+stats.StandardDeviation().ToStringMixed());
    display_graph(tick_freq,"Ticks","Frequency");
#if CALCULATE_LOOP_STATISTICS
      Console.WriteLine("Max loops: "+loop_freq.MaxFreq);
      stats.Calc(loop_freq);
      Console.WriteLine("Sample size: "+stats.Size());
      Console.WriteLine("Average: "+stats.Average().ToStringMixed());
      Console.WriteLine("Median: "+stats.Median());
      Console.WriteLine("Mode: "+stats.Mode());
      Console.WriteLine("Standard Deviation: "+stats.StandardDeviation().ToStringMixed());
      display_graph(loop_freq,"Loops","Frequency");
#else
      Console.WriteLine("\nStatistics for loop count not gathered. To enable loop statistics,\n");
      Console.WriteLine("recompile defining \"CALCULATE_LOOP_STATISTICS\"\n");
#endif
  }

  static void DoTest(int denominator,FrequencyArray tick_freq,FrequencyArray loop_freq)
  {
    Stopwatch stopWatch; // = Stopwatch.StartNew();
    Fraction f = new Fraction(0.0);
/*    stopWatch.Stop();
    stopWatch = Stopwatch.StartNew();
    f.Set(00.2);
    stopWatch.Stop();*/
    int i;
    for(i=0;i<denominator;i++) {
      stopWatch = Stopwatch.StartNew();
      f.Set(((double)i)/((double)denominator));
      stopWatch.Stop();
      // According to Microsoft docs, when using Stopwatch, the first execution should be ignored
      if(i>0) {
        tick_freq.Increment((int)stopWatch.Elapsed.Ticks);
#if CALCULATE_LOOP_STATISTICS
        loop_freq.Increment(Fraction.loops);
#endif
      }
    }
  }

  static void SingleTest(int denominator)
  {
    FrequencyArray tick_freq = new FrequencyArray();
    FrequencyArray loop_freq = new FrequencyArray();
    DoTest(denominator,tick_freq,loop_freq);
    tick_freq.Sort();
    loop_freq.Sort();
    ShowResults(tick_freq,loop_freq);
  }
  public static void Main(string[] args)
  {
/*    Fraction f = new Fraction();
    long start=DateTime.Now.Millisecond;;
    f.Set(.123456);
    long elapsed = DateTime.Now.Millisecond - start;
    Console.WriteLine(elapsed);*/
    SingleTest(1000);
  }
}

File: Pascal/fraction_performance.pas
(*
		Copyright (C) 2019-2021  by Terry N Bezue

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*)

program fraction_performance;
uses unixtype,crt,fraction,sysutils,linux,getopts;

type
  Frequency = object
  public
    value,frequency : longint;
end;

type
  Statistics = object
  avg,standard_deviation: real;
  sample_size,sample_median,sample_mode: longint;

  function Average: TMixedFraction;
  function StandardDeviation: TMixedFraction;
  function Count: longint;
  function Mode: longint;
  function Median: longint;
end;

function Statistics.Average: TMixedFraction;
begin
  Average.fset(avg);
  Average := Average.Round(100);
end;

function Statistics.StandardDeviation: TMixedFraction;
begin
  StandardDeviation.fset(standard_deviation);
  StandardDeviation := StandardDeviation.Round(100);
end;

function Statistics.Count: longint;
begin
  Count := sample_size;
end;

function Statistics.Median: longint;
begin
  Median := sample_median;
end;

function Statistics.Mode: longint;
begin
  Mode := sample_mode;
end;

type
  FrequencyArray = object
  freq_array: array of Frequency;
  max_freq: longint;

  public
    function size: longint;
    function maxFreq: longint;
    procedure increment(value: longint);
    procedure Sort;
    function Statistics: Statistics;
    procedure DisplayGraph(const xlabel,ylabel: string);
    procedure ShowResults(const heading,xlabel: string);
end;

function FrequencyArray.size: longint;
begin
  size := length(freq_array);
end;

function FrequencyArray.maxFreq: longint;
begin
  maxFreq := max_freq;
end;

procedure FrequencyArray.increment(value: longint);
var
  i: integer;
  found: boolean;

begin
  found := false;
  for i := 0 to high(freq_array) do
  begin
    if(freq_array[i].value = value) then
    begin
      found := true;
      break;
    end;
  end;
  if found = false then
  begin
    i:=length(freq_array);
    setlength(freq_array,i+1);
    freq_array[i].value := value;
    freq_array[i].frequency := 0;
  end;
  freq_array[i].frequency := freq_array[i].frequency + 1;
end;

procedure FrequencyArray.Sort;
var
  i,j: integer;
  temp_freq: Frequency;
begin
  for i := 0 to high(freq_array)-1 do
    for j := i+1 to high(freq_array) do
    begin
      if freq_array[i].value > freq_array[j].value then
      begin
        temp_freq := freq_array[i];
        freq_array[i] := freq_array[j];
        freq_array[j] := temp_freq;
      end;
    end;

end;

function FrequencyArray.Statistics: Statistics;
var
  i,sum,count: longint;
  variance: real;
begin
  sum := 0;
  Statistics.sample_size:=0;
  max_freq := 0;
  for i := 0 to high(freq_array) do
  begin
    Statistics.sample_size += freq_array[i].frequency;
    sum += freq_array[i].frequency*freq_array[i].value;
    if freq_array[i].frequency > max_freq then
    begin
      max_freq := freq_array[i].frequency;
      Statistics.sample_mode := freq_array[i].value;
    end
  end;
  Statistics.avg := sum/Statistics.sample_size;

  count := 0;
  variance := 0;
  Statistics.sample_median := -1;
  for i := 0 to high(freq_array) do
  begin
    variance += sqr(freq_array[i].value - Statistics.avg)*freq_array[i].frequency;
    if Statistics.sample_median = -1 then
    begin
      count += freq_array[i].frequency;
      if count >= trunc(Statistics.sample_size/2) then
        Statistics.sample_median := freq_array[i].value;
    end;
  end;

  Statistics.standard_deviation := sqrt(variance/(Statistics.sample_size-1));
end;

procedure FrequencyArray.DisplayGraph(const xlabel,ylabel: string);
  var
    i: longint;
    scale: real;
begin
  scale := 55/max_freq;
  writeln;
  writeln(format('%5s|              %s',[xlabel,ylabel]));
  writeln(StringOfChar('-',55));
  for i:=0 to high(freq_array) do
    writeln(format('%4d |%s %d',[freq_array[i].value,StringOfChar('#',trunc(freq_array[i].frequency*scale)),
          freq_array[i].frequency]));
  writeln;
end;

procedure FrequencyArray.ShowResults(const heading,xlabel: string);
var
  s: Statistics;

begin
  writeln;
  writeln(heading);
  writeln;
  s := Statistics;
  writeln('  Sample size: ',s.Count());
  writeln('  Min ',xlabel,': ',freq_array[0].value);
  writeln('  Max ',xlabel,': ',freq_array[size-1].value);
  writeln('  Average: ',string(s.Average));
  writeln('  Median: ',s.Median);
  writeln('  Mode: ',s.Mode);
  writeln('  Standard Deviation: ',string(s.StandardDeviation));
  DisplayGraph(xlabel,'Frequency');
end;

function time_diff_in_tens_ns(start,finish : timespec): longint;
begin
  time_diff_in_tens_ns := round(((finish.tv_sec - start.tv_sec)*1000000000 + (finish.tv_nsec - start.tv_nsec))/10.0);
end;

type PFrequencyArray = ^FrequencyArray;

procedure DoTest(value: real; var time_freq,loop_freq: PFrequencyArray);
var
  i: longint;
  f: TFraction;
  start,finish: timespec;

begin
  clock_gettime(CLOCK_PROCESS_CPUTIME_ID,@start);
  for i := 1 to 1000 do
    f.fset(value);
  clock_gettime(CLOCK_PROCESS_CPUTIME_ID,@finish);
  time_freq^.increment(round(time_diff_in_tens_ns(start,finish)/1000));
  loop_freq^.increment(loops);
end;

procedure SingleTest(denominator: longint);
var
  time_freq,loop_freq: FrequencyArray;
  ptime_freq,ploop_freq: PFrequencyArray;
  i: longint;
begin
  ptime_freq := @time_freq;
  ploop_freq := @loop_freq;
  for i := 1 to denominator-1 do
    DoTest(i/denominator,ptime_freq,ploop_freq);

  time_freq.Sort;
  loop_freq.Sort;
  time_freq.ShowResults('Time (in tens of nanoseconds) to convert floating point to fraction','Time');
  loop_freq.ShowResults('Iterations to convert floating point to fraction','Loops');
end;

procedure RandomTest(nTests: longint);
var
  i: longint;
  time_freq,loop_freq: FrequencyArray;
  ptime_freq,ploop_freq: PFrequencyArray;

begin
  Randomize;
  ptime_freq := @time_freq;
  ploop_freq := @loop_freq;
  for i:= 1 to nTests do
    DoTest(Random(1000)/Random(1000),ptime_freq,ploop_freq);
  time_freq.Sort;
  loop_freq.Sort;
  time_freq.ShowResults('Time (in tens of nanoseconds) to convert floating point to fraction','Time');
  loop_freq.ShowResults('Iterations to convert floating point to fraction','Loops');
end;

procedure syntax;
var
  pgm: string;
begin
  pgm := paramstr(0);
  writeln('Syntax: ',pgm,' [-h | --help ]\n');
  writeln('        ',pgm,' [ [-s | --single] N] [ [-r | --random] N ]');
  writeln;
  writeln('        ',pgm);
  writeln;
  writeln('Where:  -h | --help prints this help message');
  writeln('        -s | --single N -- gather statistics using N as denominator (runs tests using fractions 1/N to (N-1)/N)');
  writeln('        -r | --random N -- gather statistics running N tests using random values');
  writeln('        The default is to run a single test using 1000 as denominator and 1000 random tests');
  writeln('Examples');
  writeln('   1) To run default case');
  writeln('      ',pgm);
  writeln('   2) To run single test using denominator of 100000');
  writeln('      ',pgm,' -s 100000');
  writeln('   3) To run a minimum of 30000 random test');
  writeln('      ',pgm,' -r 30000');
  writeln('   4) To run a single test using denominator of 100000 and a minimum of 30000 random test');
  writeln('      ',pgm,' --single 100000 --random 30000');

  writeln;
  writeln;
end;

var
  needHelp : boolean;
  c: char;
  optionindex,denominator,maxTests : Longint;
  options : array[1..3] of TOption = (
      ( name:'help';has_arg:0;flag:nil;value:'h' ),
      ( name:'single';has_arg:1;flag:nil;value:'s' ),
      ( name:'random';has_arg:1;flag:nil;value:'r' )
    );

begin
  needHelp := false;
  denominator := -1;
  maxTests := -1;
  if paramcount > 0 then
  begin
    repeat
      c:=getlongopts('hs:r:',@options[1],optionindex);
      case c of
        'h','?',':' : needHelp := true;
        's' : denominator := StrToInt(optarg);
        'r' : maxTests  := StrToInt(optarg);
      end; { case }
    until c=endofoptions;
    if needHelp then
      syntax
    else
    begin
      if denominator > 0 then
        SingleTest(denominator);
      if maxTests > 0 then
        RandomTest(maxTests);
    end;
  end
  else
  begin
    SingleTest(1000);
    RandomTest(1000);
  end;
end.
